File: Objective-C/macOS/WindowController.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
Primary window controller. Checks whether the user is allowed to make purchases and whether the resource file containing the product identifiers
exists. Disables the UI when the user is not allowed to make any purchases.
*/

#import "Utilities.h"
#import "StoreObserver.h"
#import "WindowController.h"
#import "AppConfiguration.h"
#import "MainViewController.h"

@interface WindowController ()
@property (strong) Utilities *utility;
@property (strong) MainViewController *mainViewController;
@end

@implementation WindowController
#pragma mark - Window Life Cycle

- (void)windowDidLoad {
	[super windowDidLoad];
	self.utility = [[Utilities alloc] init];
	self.mainViewController = (MainViewController *)self.window.contentViewController;

	// First, let's check whether the user is allowed to make purchases. Proceed if they are allowed. Display an alert, otherwise.
	if ([StoreObserver sharedInstance].isAuthorizedForPayments) {
		NSArray *identifiers = self.utility.identifiers;
		NSString *message;

		// Refresh the UI if the resource file containing the product identifiers exists. Show a message, otherwise.
		if (identifiers != nil) {
			if (identifiers.count > 0) {
				[self.mainViewController reloadViewController:PCSViewControllerNamesProducts];
			} else {
				message = [NSString stringWithFormat:@"%@.%@ %@", PCSProductIdsPlistName, PCSProductIdsPlistFileExtension, PCSMessagesEmptyResource];
				// Warn the user that the resource file does not contain anything.
				[self.mainViewController reloadViewController:PCSViewControllerNamesMessages withMessage:message];
			}
		} else {
			message = [NSString stringWithFormat:@"%@ %@.%@.", PCSMessagesResourceNotFound, PCSProductIdsPlistName, PCSProductIdsPlistFileExtension];
			// Warn the user that the resource file could not be found.
			[self.mainViewController reloadViewController:PCSViewControllerNamesMessages withMessage:message];
		}
	} else {
		// Warn the user that they are not allowed to make purchases.
		[self.mainViewController reloadViewController:PCSViewControllerNamesMessages withMessage:PCSMessagesCannotMakePayments];
	}
}

#pragma mark - Switches Between Products and Purchases Panes

- (IBAction)showProducts:(NSToolbarItem *)sender {
	[self.mainViewController reloadViewController:PCSViewControllerNamesProducts];
}


- (IBAction)showPurchases:(NSToolbarItem *)sender {
	[self.mainViewController reloadViewController:PCSViewControllerNamesPurchases];
}


#pragma mark - NSUserInterfaceValidations

- (BOOL)validateUserInterfaceItem:(nonnull id<NSValidatedUserInterfaceItem>)item {
	if (item.action == @selector(showProducts:)) {
		return ([StoreObserver sharedInstance].isAuthorizedForPayments && (self.utility.identifiers != nil && self.utility.identifiers.count > 0));
	} else if (item.action == @selector(showPurchases:)) {
		return [StoreObserver sharedInstance].isAuthorizedForPayments;
	}
	return NO;
}

@end

File: Swift/macOS/WindowController.swift
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
Primary window controller. Checks whether the user is allowed to make purchases and whether the resource file containing the product identifiers
exists. Disables the UI when the user is not allowed to make any purchases.
*/

import Cocoa

class WindowController: NSWindowController, NSUserInterfaceValidations {
	// MARK: - Properties

	fileprivate var resourceFile = ProductIdentifiers()
	fileprivate var mainViewController: MainViewController!

	// MARK: - Window Life Cycle

	override func windowDidLoad() {
		super.windowDidLoad()

		guard let window = window else { fatalError("\(Messages.windowDoesNotExist)") }
		guard let viewController = window.contentViewController as? MainViewController else { fatalError("\(Messages.viewControllerDoesNotExist)") }

		mainViewController = viewController

		// First, let's check whether the user is allowed to make purchases. Proceed if they are allowed. Display an alert, otherwise.
		if StoreObserver.shared.isAuthorizedForPayments {
			// Refresh the UI if the resource file containing the product identifiers exists. Show a message, otherwise.
			guard let identifiers = resourceFile.identifiers else {
				// Warn the user that the resource file could not be found.
				mainViewController.reloadViewController(.messages, with: "\(resourceFile.wasNotFound)")
				return
			}

			// Refresh the UI if the resource file containing the product identifiers exists. Show a message, otherwise.
			if !identifiers.isEmpty {
				mainViewController.reloadViewController(.products)
			} else {
				// Warn the user that the resource file does not contain anything.
				mainViewController.reloadViewController(.messages, with: "\(resourceFile.isEmpty)")
			}
		} else {
			// Warn the user that they are not allowed to make purchases.
			mainViewController.reloadViewController(.messages, with: Messages.cannotMakePayments)
		}
	}

	// MARK: - Switches Between Products and Purchases Panes

	@IBAction fileprivate func showProducts(_ sender: NSToolbarItem) {
		guard ViewControllerNames(rawValue: sender.label) != nil else { fatalError("\(Messages.unknownToolbarItem)\(sender.label).") }
		mainViewController.reloadViewController(.products)
	}

	@IBAction func showPurchases(_ sender: NSToolbarItem) {
		guard ViewControllerNames(rawValue: sender.label) != nil else { fatalError("\(Messages.unknownToolbarItem)\(sender.label).") }
		mainViewController.reloadViewController(.purchases)
	}

	// MARK: - NSUserInterfaceValidations

	func validateUserInterfaceItem(_ item: NSValidatedUserInterfaceItem) -> Bool {
		let identifiers = resourceFile.identifiers

		if item.action == #selector(WindowController.showProducts(_:)) {
			return StoreObserver.shared.isAuthorizedForPayments && (identifiers != nil && !(identifiers!.isEmpty))
		} else if item.action == #selector(WindowController.showPurchases(_:)) {
			return StoreObserver.shared.isAuthorizedForPayments
		}
		return false
	}
}
